fix #285040 crash when changing a triplet's rest's duration

WcRnm opened the pull request:
Resolves: https://musescore.org/en/node/285040

Prevents a null pointer dereference and subsequent app crash.

- [x] I signed [CLA](https://musescore.org/en/cla)
- [x] I made sure the code in the PR follows [the coding rules](https://musescore.org/en/handbook/developers-handbook/finding-your-way-around/musescore-coding-rules)
- [x] I made sure the code compiles on my machine
- [x] I made sure there are no unnecessary changes in the code
- [x] I made sure the title of the PR reflects the core meaning of the issue you are solving
- [x] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"
- [x] I made sure the commit message title starts with "fix #424242:" if there is a related issue
- [ ] I created the test (mtest, vtest, script test) to verify the changes I made

mattmcclinch commented:
Why did you rearrange so much code?
```
newcr->setTuplet(nt);
nt->setParent(newcr->measure());
```
is now dependent on the condition on line 4746 being false, and
```
qWarning("linked tuplet not found");
```
is no longer executed if `le` is null.

I think that in its current form, this PR has the potential to cause more problems than it solves, especially since the original problem depends on the score being corrupt in the first place. It is unclear to me what should happen after we realize we can't find the linked tuplet, but if the operation is allowed to succeed, it will only cause more discrepancies between the master score and the part.

Jojo-Schmitz commented:
Under which name (on musescore.org) did you sign the CLA?

WcRnm commented:
@Jojo-Schmitz 

I signed the CLA as "Rory McLeod McLeod". I put in a request to get fixed - it should be "Rory Neil McLeod".

When I added a comment to https://musescore.org/en/node/285040 noting this PR, I changed the status to "PR Created". I did get a notice that the post was pending review. 

Should I leave status fields unchanged when commenting on a particular issue? I'm fine leaving those decisions up to the moderators and more experienced developers on this project.

When ready, I will resubmit with the corrected comment. Is a forced push okay, or is a clean PR the preferred method?

------------------------------------------------------
@mattmcclinch 
Thank you for the feedback. My change does indeed have problems. The code uses indentation formatting that I am not used to - this in entirely my issue, and I should have been more careful.

Looking back at the original code the cause of the crash is on line 4804-4805. In the test case `n` 'is null. So simply changing the code from:
```
newcr->setTuplet(nt);
nt->setParent(newcr->measure());
```
to:
```
if (nt != 0) {
      newcr->setTuplet(nt);
      nt->setParent(newcr->measure());
      }
```
is sufficient to prevent the crash, also, notably, the expected behavior is also fixed, however the missing tuplet warning message is still hit repeatedly when the rest duration is changed.

IMO, there are two separate issues here. 
1) The app crash, which this PR is intended to fix.
2) An issue that occurs when reading in a score and results in the "out of sync" state for the tuplet as noted in the issue comments.

mattmcclinch commented:
So you are [mcleodnw](https://musescore.org/en/user/2345918), the latest person to have signed the CLA.

You did the right thing by changing the status to "PR created". I have no idea what you may have done to trigger the spam prevention mechanism.

I am much happier with the fix that you are proposing now, though we usually write
```
if (nt)
```
rather than
```
if (nt != 0)
```

Forced pushes are okay; no need for a clean PR.

Howard-C commented:
Better change the title to "fix #285040:" rather than "fix: #285040" so [#285040](https://musescore.org/node/285040) can get closed automatically after the PR is merged.

mattmcclinch commented:
Simply changing the title won't work. You have to change the commit message.

WcRnm commented:
I fixed both the commit message and PR title.

Jojo-Schmitz commented:
> When I added a comment to https://musescore.org/en/node/285040 noting this PR, I changed the status to "PR Created". I did get a notice that the post was pending review.

Ah, I see, guess your post contained a link that musescore.org deemed spammy

WcRnm commented:
Before this PR is merged, I should try my hand at writing a unit test.

mattmcclinch commented:
We don't need to add another test every time we check for null before dereferencing a pointer.

WcRnm commented:
I edited the score down to just the one isolated measure and it still hits the warning message. The main score has an initial half note rest on voice 1, whereas the electric guitar score has a quarter note.

I will add a comment to the associated issue about this.

![main-view](https://user-images.githubusercontent.com/3332121/67031324-1a7b7480-f0c6-11e9-99c1-ef68d154ec2c.png)
![electric-guitar-view](https://user-images.githubusercontent.com/3332121/67031331-1cddce80-f0c6-11e9-9adf-4b33c49a8154.png)

mattmcclinch commented:
Yes, that is (part of) what I meant when I said the parts got out of sync.

RobFog commented:
https://musescore.org/en/node/285040 is marked as "P0 - Critical". Can this be added to the 3.4 milestone?
